File: 100289/ko-papir-ollo/Kopapirollo.java
import java.util.Scanner;
import java.util.TreeMap;
import java.util.ArrayList;

class Kopapirollo {
  static TreeMap<String, ArrayList<String>> nyertesek = new TreeMap<String, ArrayList<String>>();
  public static void main(String[] args) {
    Kopapirollo.megold();
  }

  public static void megold() {
    Scanner scanner = new Scanner(System.in);
    while(scanner.hasNext()) {
      String[] sor = scanner.nextLine().split(";");
      String elsoJatekos = sor[0];
      String elsoJel = sor[1];
      String masodikJatekos = sor[2];
      String masodikJel = sor[3];
      String[] kiertekeles = Kopapirollo.kiertekel(elsoJatekos, elsoJel, masodikJatekos, masodikJel);
      if (kiertekeles != null) {
        beletesz(kiertekeles[0], kiertekeles[1]);
      }
    }
    scanner.close();
    kiirat();
  }

  public static String[] kiertekel(String elsoJatekos, String elsoJel, String masodikJatekos, String masodikJel) {
    if (elsoJel.equals(masodikJel)) {
      return null;
    }
    if (elsoJel.equals("ko")) {
      return masodikJel.equals("papir") ? new String[]{masodikJatekos, masodikJel} : new String[]{elsoJatekos, elsoJel};
    }
  
    if (elsoJel.equals("papir")) {
      return masodikJel.equals("ollo") ? new String[]{masodikJatekos, masodikJel} : new String[]{elsoJatekos, elsoJel};
    }

    // if ollo
    return masodikJel.equals("ko") ? new String[]{masodikJatekos, masodikJel} : new String[]{elsoJatekos, elsoJel};
  }

  public static void beletesz(String jatekos, String jel) {
    ArrayList<String> jatekosEddigiNyeresei = nyertesek.get(jatekos);
    if (jatekosEddigiNyeresei == null) {
      ArrayList<String> kezdetiLista = new ArrayList<String>();
      kezdetiLista.add(jel);
      nyertesek.put(jatekos, kezdetiLista); // ez itt bug-os
    } else {
      jatekosEddigiNyeresei.add(jel);
    }
  }

  public static void kiirat() {
    nyertesek.forEach((jatekos, jelek) -> {
      String kiirandoSzoveg = String.format("%s: %s", jatekos, String.join(", ", jelek));
      System.out.println(kiirandoSzoveg);
    });
  }

}

File: 100289/ko-papir-ollo/Kopapirollo.cs
using System;
using System.Collections.Generic;

public class Program {
  static SortedDictionary<string, List<string>> nyertesek = new SortedDictionary<string, List<string>>();
  public static void Main()
	{
		megold();
	}

  public static void megold() {
    string Line;
    
    while( !String.IsNullOrEmpty( (Line = Console.ReadLine()) ) ) {
      string[] sor = Line.Split(';');
      string elsoJatekos = sor[0];
      string elsoJel = sor[1];
      string masodikJatekos = sor[2];
      string masodikJel = sor[3];
      string[] kiertekeles = kiertekel(elsoJatekos, elsoJel, masodikJatekos, masodikJel);
      if (kiertekeles != null) {
        beletesz(kiertekeles[0], kiertekeles[1]);
      }
    }
    kiirat();
  }

  public static string[] kiertekel(string elsoJatekos, string elsoJel, string masodikJatekos, string masodikJel) {
    if (elsoJel.Equals(masodikJel)) {
      return null;
    }
    if (elsoJel.Equals("ko")) {
      return masodikJel.Equals("papir") ? new string[]{masodikJatekos, masodikJel} : new string[]{elsoJatekos, elsoJel};
    }
  
    if (elsoJel.Equals("papir")) {
      return masodikJel.Equals("ollo") ? new string[]{masodikJatekos, masodikJel} : new string[]{elsoJatekos, elsoJel};
    }

    // if ollo
    return masodikJel.Equals("ko") ? new string[]{masodikJatekos, masodikJel} : new string[]{elsoJatekos, elsoJel};
  }

  public static void beletesz(string jatekos, string jel) {
    if(nyertesek.ContainsKey(jatekos)) {
      nyertesek[jatekos].Add(jel);
    } else {
      List<string> ujJatekosListaja = new List<string>();
      ujJatekosListaja.Add(jel);
      nyertesek.Add(jatekos, ujJatekosListaja);
    }
  }

  public static void kiirat() {
    foreach(var item in nyertesek) {
      string kiirandoSzoveg = string.Format("{0}: {1}", item.Key, string.Join(", ", item.Value));
      Console.WriteLine(kiirandoSzoveg);
    }
  }

}